### PR TITLE
Api refacto

### PR DESCRIPTION
**Pull request by robechon**
It is now possible to do POST request with request method

### CODE REVIEW

**Comment by fwininger** [comment]
Can you rebase please.

File: cbw_api_toolbox/cbw_api.py [hunk 1]
```diff
@@ -67,11 +74,15 @@ def ping(self):
...
         """GET request to /api/v2/servers to get all servers"""
-        return [CBWParser().parse(CBWServer, server) for server in self._request(ROUTE_GET_SERVERS)]
+        return [CBWParser().parse(CBWServer, server) for server in self._request(
+            "GET",
+            self._build_route(ROUTE_SERVERS))]
 
     def server(self, server_id):
         """GET request to /api/v2/servers to get all informations about a specific server"""
-        return CBWParser().parse(CBWServer, self._request(ROUTE_GET_SERVERS, [server_id]))
+        return CBWParser().parse(CBWServer, self._request(
```
Comment (fwininger) [inline cbw_api_toolbox/cbw_api.py]: Why 3 lines ?

File: cbw_api_toolbox/cbw_api.py [hunk 2]
```diff
@@ -67,11 +74,15 @@ def ping(self):
 
     def servers(self):
         """GET request to /api/v2/servers to get all servers"""
-        return [CBWParser().parse(CBWServer, server) for server in self._request(ROUTE_GET_SERVERS)]
+        return [CBWParser().parse(CBWServer, server) for server in self._request(
```
Comment (fwininger) [inline cbw_api_toolbox/cbw_api.py]: Why 3 lines ?

File: cbw_api_toolbox/cbw_api.py [hunk 3]
```diff
@@ -31,32 +31,39 @@ def __init__(self, api_url, api_key, secret_key, verify_ssl=False):
...
+                data=body_params,
+                auth=CBWAuth(self.api_key, self.secret_key),
+                verify=self.verify_ssl)
             return json.loads(response.text)
 
         except JSONDecodeError as error:
-            self.logger.error("An error occurred when decoding {0} with route {1}/{2}{3} : {4}"
-                              .format(response.text, self.api_url, route,
-                                      self._get_params(params), error))
+            self.logger.error("An error occurred when decoding %s with route %s : %s",
```
Comment (fwininger) [inline cbw_api_toolbox/cbw_api.py]: Please use .format

File: cbw_api_toolbox/cbw_api.py [hunk 4]
```diff
@@ -31,32 +31,39 @@ def __init__(self, api_url, api_key, secret_key, verify_ssl=False):
     def _get_params(params=None):
```
Comment (fwininger) [inline cbw_api_toolbox/cbw_api.py]: Remove this method

File: cbw_api_toolbox/cbw_api.py [hunk 5]
```diff
@@ -31,32 +31,39 @@ def __init__(self, api_url, api_key, secret_key, verify_ssl=False):
...
         return "/{0}".format('/'.join(params)) if params else ""
 
-    def _request(self, route, params=None):
+    def _build_route(self, route, params=None):
```
Comment (fwininger) [inline cbw_api_toolbox/cbw_api.py]: Use this method only in _request

File: cbw_api_toolbox/cbw_api.py [hunk 6]
```diff
@@ -31,32 +31,39 @@ def __init__(self, api_url, api_key, secret_key, verify_ssl=False):
...
+        return "{0}{1}{2}".format(self.api_url, route, self._get_params(params))
+
+    def _request(self, verb, route, body_params=None):
         response = None
+
+        if body_params is not None:
+            body_params = json.dumps(body_params)
```
Comment (fwininger) [inline cbw_api_toolbox/cbw_api.py]: Add one line after.

File: cbw_api_toolbox/cbw_api.py (hunk 2 above)
Comment (robechon) [reply cbw_api_toolbox/cbw_api.py]: Pylint doesn't want to let one line, too many characters ... (Pylint message: "Line too long 122/100")

File: cbw_api_toolbox/cbw_api.py (hunk 1 above)
Comment (robechon) [reply cbw_api_toolbox/cbw_api.py]: Pylint doesn't want to let one line, too many characters ... (Pylint message: "Line too long 116/100")

File: cbw_api_toolbox/cbw_api.py (hunk 3 above)
Comment (robechon) [reply cbw_api_toolbox/cbw_api.py]: Pylint doesn't let me use format (Pylint message: "Use % formatting in logging functions and pass the % parameters as arguments")

File: cbw_api_toolbox/cbw_api.py (hunk 4 above)
Comment (robechon) [reply cbw_api_toolbox/cbw_api.py]: Done !

File: cbw_api_toolbox/cbw_api.py (hunk 6 above)
Comment (robechon) [reply cbw_api_toolbox/cbw_api.py]: Done !

File: cbw_api_toolbox/cbw_api.py (hunk 5 above)
Comment (robechon) [reply cbw_api_toolbox/cbw_api.py]: Done !

File: cbw_api_toolbox/cbw_api.py (hunk 2 above)
Comment (fwininger) [reply cbw_api_toolbox/cbw_api.py]: Mettre 

```python
  for server in self._request("GET", [ROUTE_SERVERS])] : 
     CBWParser().parse(CBWServer, server)
```

File: cbw_api_toolbox/cbw_api.py [hunk 7]
```diff
@@ -27,36 +27,41 @@ def __init__(self, api_url, api_key, secret_key, verify_ssl=False):
         self.verify_ssl = verify_ssl
         self.logger = logging.getLogger(self.__class__.__name__)
 
-    @staticmethod
-    def _get_params(params=None):
-        return "/{0}".format('/'.join(params)) if params else ""
+    def _build_route(self, params=None):
+        params = "{0}".format('/'.join(params))
+        return "{0}{1}".format(self.api_url, params)
```
Comment (fwininger) [inline cbw_api_toolbox/cbw_api.py]:
```python
return "{0}{1}".format(self.api_url, '/'.join(params))
```

File: cbw_api_toolbox/cbw_api.py [hunk 8]
```diff
@@ -27,36 +27,41 @@ def __init__(self, api_url, api_key, secret_key, verify_ssl=False):
         self.verify_ssl = verify_ssl
         self.logger = logging.getLogger(self.__class__.__name__)
 
-    @staticmethod
-    def _get_params(params=None):
-        return "/{0}".format('/'.join(params)) if params else ""
+    def _build_route(self, params=None):
```
Comment (fwininger) [inline cbw_api_toolbox/cbw_api.py]: params=None do work.

File: cbw_api_toolbox/cbw_api.py [hunk 9]
```diff
@@ -27,36 +27,40 @@ def __init__(self, api_url, api_key, secret_key, verify_ssl=False):
...
+                data=body_params,
+                auth=CBWAuth(self.api_key, self.secret_key),
+                verify=self.verify_ssl)
             return json.loads(response.text)
 
         except JSONDecodeError as error:
-            self.logger.error("An error occurred when decoding {0} with route {1}/{2}{3} : {4}"
-                              .format(response.text, self.api_url, route,
-                                      self._get_params(params), error))
+            self.logger.error("An error occurred when decoding {0} with route {1} : {2}".format(
```
Comment (jinxka) [inline cbw_api_toolbox/cbw_api.py]: Utiliser logger.exception

https://docs.python.org/3/library/logging.html

File: cbw_api_toolbox/cbw_api.py [hunk 10]
```diff
@@ -27,36 +27,40 @@ def __init__(self, api_url, api_key, secret_key, verify_ssl=False):
...
+                response.text, route, error))
 
         except (ConnectionError, ProxyError, SSLError, NewConnectionError, RetryError,
                 InvalidHeader, MaxRetryError) as error:
-            self.logger.error("An error occurred when requesting {0}/{1}{2} : {3}"
-                              .format(self.api_url, route, self._get_params(params), error))
+            self.logger.error("An error occurred when requesting {0} : {1}".format(route, error))
```
Comment (jinxka) [inline cbw_api_toolbox/cbw_api.py]: Utiliser logger.exception

https://docs.python.org/3/library/logging.html

**Comment by jinxka** [comment]
OK

**Pull request merged**